# Business Understanding


In [ ]:
# Importando librerias
import pandas as pd
import numpy as np
import matplotlib.pyplot as plt
import pickle as pke

In [ ]:
from google.colab import drive
drive.mount('/content/drive')

KeyboardInterrupt: ignored

In [ ]:
from pandas.io.pickle import read_pickle
import pandas as pd
df = open('/content/drive/MyDrive/Machine learnig/Caso semestral/senate2020.pkl','rb')
data = pd.read_pickle(df)
data

0    \
data     Latorre R., Juan Ignacio                                                        Si   
         Coloma C., Juan Antonio                                                         Si   
         García R., José                                                                 Si   
         Bianchi C., Carlos                                                              Si   
         Navarro B., Alejandro                                                           Si   
         Allende B., Isabel                                                              Si   
         Lagos W., Ricardo                                                               Si   
         García Huidobro S., Alejandro                                                   Si   
         Guillier Á., Alejandro                                                       Pareo   
         Durana S., José Miguel                                                          Si   
         Prohens E., Rafael                                                              Si   
         Pugh O., Kenneth                                                                Si   
         Elizalde S., Alvaro                                                             Si   
         Galilea V., Rodrigo                                                             Si   
         Huenchumilla J., Francisco                                                      Si   
         Sandoval P., David                                                              Si   
         Pizarro S., Jorge                                                              NaN   
         Girardi L., Guido                                                              NaN   
         Letelier M., Juan Pablo                                                        NaN   
         Pérez V., Víctor                                                               NaN   
         Chahuán C., Francisco                                                          NaN   
         Quintana L., Jaime                                                             NaN   
         Rincón G., Ximena                                                              NaN   
         Von Baer J., Ena                                                               NaN   
         Araya G., Pedro                                                                NaN   
         Van Rysselberghe H., Jacqueline                                                NaN   
         Goic B., Carolina                                                              NaN   
         Harboe B., Felipe                                                              NaN   
         De Urresti L., Alfonso                                                         NaN   
         Quinteros L., Rabindranath                                                     NaN   
         Moreira B., Iván                                                               NaN   
         Muñoz D., Adriana                                                              NaN   
         Ossandón I., Manuel José                                                       NaN   
         Insulza S., José Miguel                                                        NaN   
         Ebensperger O., Luz Eliana                                                     NaN   
         Soria Q., Jorge                                                                NaN   
         Provoste C., Yasna                                                             NaN   
         Castro P., Juan                                                                NaN   
         Aravena A., Carmen Gloria                                                      NaN   
         Kast S., Felipe                                                                NaN   
         Órdenes N., Ximena                                                             NaN   
         Allamand Z., Andrés                                                            NaN   
   

## Contexto

El caso se trata de un listado de la cámara de senadores el cual para poder leer este tuvimos que investigar la manera de hacerlo en la cual dimos con el objeto pickle el cual nos facilita la visualización de los datos de este el cual está dividido en dos  partes.

Data:En el data podemos encontrar datos de  cada uno de los senadores con sus respectivos nombres,apellidos y los días que estos estuvieron presentes en la cámara de senadores

MetaData:En el MetaData podemos encontrar datos como la “etapa” en la cual se encuentra la votación la cual comúnmente va desde el primer trámite hasta el tercer trámite también tenemos la “fecha” en la cual se realizó la votación,el “tema” en ese dato podemos apreciar el tema que se va a tratar durante esta votación,el “quórum”


Otro tipos de datos que tenemos es la decisión que tomó el senador a la hora de votar en la cual tenemos el “si” este dato representa la cantidad total de senadores que estuvieron a favor por lo cual se estaba votando durante esa etapa del proceso, el “no” este dato representa la cantidad total de senadores que estuvieron en contra por lo cual se estaba votando en esa etapa del proceso.

la “abstención”este dato representa la cantidad total de senadores que tomaron la decisión de no votar es decir de guardarse su voto, el ”pareo” este dato representa la cantidad total de senadores que realizaron este tratado el cual consiste en que dos senadores se ponen de acuerdo para realizar el mismo voto en el caso que de uno de estos dos senadores no se encuentre presente el otro senador no podrá ejercer su voto.


Todo esto se encuentra en dataframe qué tiene un listado qué corresponde a la informaciòn de la cámara de senadores qué contienen el nombre del senador, el voto del senador, el tipo de voto, conteo de los diferentes votos, nombre del proyecto, etapa del proyecto, tipo del resultado, fecha y hora a la qué inició la sesión del senado.




# Data understanding

## Tipos de variables

**Cuantitativas**

Abstención.- Cuantitativa

Fecha.- Cuantitativa

No.- Cuantitativa

Pareo.- Cuantitativa

Sesión.- Cuantitativa

Si.- Cuantitativa

**Cualitativas**

Etapa.- Ordinal

Quórum.- Nominal

Tema.- Nominal

Tipo de votación.-  Nominal


In [ ]:
from pandas.io.pickle import read_pickle
df = open('/content/drive/MyDrive/Machine learnig/Caso semestral/senate2020.pkl','rb')
data = pke.load(df)
data.isnull().sum().sum()

5894

## Datos perdidos

Para lograr hacer un mejor analisis de esto se deben pasar los objetos a numeros y usar el multindex metadata

Tenemos muchos dato perdidos en las deciciones de los senadores a la hora de votar en un etapa esto puede ser que se debe a que el senador no se presento a la sala ese dia por lo cual no ejercio su voto y este dato es dato perdido o nulo

## Datos atípicos

Los datos atipicos son datos pocos comunes por lo cual dentro de esta data podemos encontrar que el dato menos comun es el pareo debido a que muy pocas veces un senador realiza un acuerdo con el otro para poder realizar su voto en conjunto.

Otro dato atipico que podemos encontar es que en la seccion del dia 2019-04-17
se presentaron solo 3 senadores en la sala lo cual sale de lo comun debido a que normalmente se tiene presente una cantidad de gente de 20 senadores en adelante esto puede ser debido a que en esa etapa se dio un nuevo plazo para indicaciones, en el cual se dio plazo para presentar indicaciones hasta el 06/05/2019.

## Correlación de variables.
Los datos de si, no, pareo y abstención están altamente relacionados con el Quórum ya qué todos estos datos son los resultados de las votaciones realizadas por los senadores el cual el qurum pide una cantidad minima de senadores para poder realizar la votación



# Data preparation

##Preparando data frame para trabajar

###Data

In [ ]:
data.unstack()
solodata = data.loc['data']
solodata

0    1           2    3    4    5    6    \
Latorre R., Juan Ignacio            Si   No          Si   Si   Si   Si   Si   
Coloma C., Juan Antonio             Si  NaN          Si   Si   Si   Si   Si   
García R., José                     Si   Si          Si   Si   Si   Si   Si   
Bianchi C., Carlos                  Si   Si          Si   Si   Si   Si   Si   
Navarro B., Alejandro               Si   Si         NaN   Si   Si   Si   Si   
Allende B., Isabel                  Si   Si          Si  NaN   Si   Si   Si   
Lagos W., Ricardo                   Si  NaN          Si  NaN   Si  NaN   Si   
García Huidobro S., Alejandro       Si  NaN          Si   Si   Si   Si   Si   
Guillier Á., Alejandro           Pareo   Si          Si   Si   Si   Si   Si   
Durana S., José Miguel              Si   Si         NaN   Si   Si   Si   Si   
Prohens E., Rafael                  Si   Si          Si   Si  NaN   Si   Si   
Pugh O., Kenneth                    Si   Si         NaN   Si   Si   Si   Si   
Elizalde S., Alvaro                 Si   Si          Si   Si   Si   Si   Si   
Galilea V., Rodrigo                 Si   Si       Pareo   Si   Si   Si  NaN   
Huenchumilla J., Francisco          Si   Si          Si   Si   Si   Si   Si   
Sandoval P., David                  Si   Si          Si   Si   Si   Si   Si   
Pizarro S., Jorge                  NaN   Si          Si   Si   Si   Si   Si   
Girardi L., Guido                  NaN   Si          Si   Si  NaN   Si   Si   
Letelier M., Juan Pablo            NaN   Si          Si   Si   Si   Si   Si   
Pérez V., Víctor                   NaN   Si         NaN   Si   Si  NaN  NaN   
Chahuán C., Francisco              NaN   Si          Si   Si   Si   Si   Si   
Quintana L., Jaime                 NaN   Si          Si   Si  NaN   Si   Si   
Rincón G., Ximena                  NaN   Si         NaN   Si   Si   Si   Si   
Von Baer J., Ena                   NaN   Si          Si   Si  NaN   Si   Si   
Araya G., Pedro                    NaN   Si         NaN  NaN  NaN   Si   Si   
Van Rysselberghe H., Jacqueline    NaN   Si  Abstencion  NaN  NaN  NaN  NaN   
Goic B., Carolina                  NaN   Si         NaN   Si   Si   Si   Si   
Harboe B., Felipe                  NaN   Si         NaN   Si   Si   Si   Si   
De Urresti L., Alfonso             NaN   Si          Si   Si   Si   Si   Si   
Quinteros L., Rabindranath         NaN   Si          Si   Si  NaN   Si   Si   
Moreira B., Iván                   NaN   Si         NaN   Si   Si   Si   Si   
Muñoz D., Adriana                  NaN   Si         NaN   Si   Si   Si   Si   
Ossandón I., Manuel José           NaN   Si          Si   Si   Si   Si   Si   
Insulza S., José Miguel            NaN   Si         NaN   Si  NaN   Si   Si   
Ebensperger O., Luz Eliana         NaN   Si          Si   Si   Si   Si   Si   
Soria Q., Jorge                    NaN   Si          Si   Si   Si  NaN   Si   
Provoste C., Yasna                 NaN   Si         NaN   Si   Si   Si  NaN   
Castro P., Juan                    NaN   Si         NaN   Si   Si   Si   Si   
Aravena A., Carmen Gloria          NaN   Si          Si   Si   Si   Si   Si   
Kast S., Felipe                    NaN   Si          Si   Si   Si  NaN  NaN   
Órdenes N., Ximena                 NaN   Si         NaN   Si   Si   Si   Si   
Allamand Z., Andrés                NaN  NaN          Si   Si   Si  NaN  NaN   
Montes C., Carlos                  NaN  NaN          Si   Si   Si  NaN   Si   
Alvarado A., Claudio               NaN  NaN         NaN  NaN  NaN  NaN   Si   
Sabat F., Marcela                  NaN  NaN         NaN  NaN  NaN  NaN   Si   

                                 7      8      9    ...    415    416    417  \
Latorre R., Juan Ignacio          Si     Si     Si  ...     Si     Si     No   
Coloma C., Juan Antonio          NaN     Si     Si  ...     Si    NaN     Si   
García R., José                   Si    NaN     Si  ...     Si    NaN     Si   
Bianchi C., Carlos                Si     Si    NaN  ...     Si     Si    NaN   


###Metadata

In [ ]:
data.unstack()
metadata = data.loc['metadata']
metadata = metadata.reset_index()
metadata = metadata.T
metadata

0                               1                    2   3  \
index  ABSTENCION                           ETAPA                FECHA  NO   
0               0   Primer trámite constitucional  2018-12-12 00:00:00   0   
1               0   Primer trámite constitucional  2020-06-02 00:00:00   1   
2               1   Primer trámite constitucional  2019-12-30 00:00:00   0   
3               0   Primer trámite constitucional  2019-03-05 00:00:00   0   
...           ...                             ...                  ...  ..   
420             4   Primer trámite constitucional  2020-09-01 00:00:00   0   
421             3  Segundo trámite constitucional  2020-09-08 00:00:00   0   
422             4  Segundo trámite constitucional  2019-04-10 00:00:00   7   
423             1  Segundo trámite constitucional  2019-06-11 00:00:00   4   
424             1  Segundo trámite constitucional  2019-06-11 00:00:00   2   

           4                     5       6   7  \
index  PAREO                QUORUM  SESION  SI   
0          1        Mayoría simple  77/366  15   
1          0                  None  31/368  37   
2          1  Cuatro séptimos Q.C.  89/367  27   
3          0        Mayoría simple  94/366  39   
...      ...                   ...     ...  ..   
420        0                  None  79/368  33   
421        0        Mayoría simple  84/368  35   
422        0        Mayoría simple   9/367  25   
423        0        Mayoría simple  24/367  33   
424        0        Mayoría simple  24/367  24   

                                                       8                     9  
index                                               TEMA          TIPOVOTACION  
0      Proyecto de ley, en primer trámite constitucio...     Discusión general  
1      Aprobación en particular del proyecto de ley q...     Discusión general  
2      Proyecto de ley, en primer trámite constitucio...     Discusión general  
3      Proyecto de ley, en primer trámite constitucio...     Discusión general  
...                                                  ...                   ...  
420    Proyecto de ley, iniciado en Moción de los Hon...     Discusión general  
421    Aprobación, en general y particular, del proye...     Discusión general  
422    Proyecto de ley, en segundo trámite constituci...     Discusión general  
423    Proyecto de ley, en segundo trámite constituci...  Discusión particular  
424    Enmienda de la Comisión de Trabajo y Previsión...  Discusión particular  

[425 rows x 10 columns]

Tras armar nuestro dataframe se detecto que la información esta almacenada en un multi-index, el primer index se llama data y el segundo metadata.

Ambos index tienen información que se complementa una con la otra pero uno cuenta la información y el otro da la información por sectores.

El metadata almacena la informacion total de las votaciones por sesion, esto se detalla mostrando las votaciones totales de la sesion, el tipo de votación, el numero de la sesion, el tema y el resultado.

El data almacena la información de los senadores, contiene cada votacion de los senadores en las diferentes sesiones y  las votaciones pueden ser Si, no, pareo o abstención.

Para trabajar en nuestro proyecto usaremos solo el metadata ya que queremos predecir cuantos Si pueden haber en una sesion dependiendo de los votos, el tipo de votacion y la etapa.

##Creando nuevo dataframe

In [ ]:
dft = pd.DataFrame(columns= ['Ab','No','TipoVotacion','Etapa','Si'])
dft['Ab']=metadata[0]
dft['No']=metadata[3]
dft['Si']=metadata[7]
dft['TipoVotacion']=metadata[9]
dft['Etapa']=metadata[1]
dft=dft.drop(dft.index[[0]])

dft

Ab No          TipoVotacion                           Etapa  Si
0    0  0     Discusión general   Primer trámite constitucional  15
1    0  1     Discusión general   Primer trámite constitucional  37
2    1  0     Discusión general   Primer trámite constitucional  27
3    0  0     Discusión general   Primer trámite constitucional  39
4    0  0  Discusión particular   Primer trámite constitucional  35
..  .. ..                   ...                             ...  ..
420  4  0     Discusión general   Primer trámite constitucional  33
421  3  0     Discusión general  Segundo trámite constitucional  35
422  4  7     Discusión general  Segundo trámite constitucional  25
423  1  4  Discusión particular  Segundo trámite constitucional  33
424  1  2  Discusión particular  Segundo trámite constitucional  24

[424 rows x 5 columns]

Aquí estamos sacando las votaciones totales de las diferentes sesiones registradas hasta la fecha exceptuando los pareo ya que cuentan como 2 votos y podría confundirse al momento de aplicarlo

##Transformacion de varibles

In [ ]:
from sklearn.preprocessing import LabelEncoder
Senado = LabelEncoder()
dft['Etapa'] = Senado.fit_transform(dft['Etapa'])
dft['TipoVotacion'] = Senado.fit_transform(dft['TipoVotacion'])
dft

Ab No  TipoVotacion  Etapa  Si
0    0  0             0      4  15
1    0  1             0      4  37
2    1  0             0      4  27
3    0  0             0      4  39
4    0  0             2      4  35
..  .. ..           ...    ...  ..
420  4  0             0      4  33
421  3  0             0      5  35
422  4  7             0      5  25
423  1  4             2      5  33
424  1  2             2      5  24

[424 rows x 5 columns]

En este caso solo transformamos la etapa y el tipo de votacion con el labelencoder, esto hace que los tipo String cambien a un int, con esto podemos someter al dataframe a un entrenamiento para implemantarlos en los modelos.

##Training

In [ ]:
import numpy as np
from sklearn.model_selection import train_test_split
X = np.array(dft.drop(['Si'],1))
y = np.array(dft['Si'])
validation_size = 0.20
seed = 7
X_train, X_test, y_train, y_test = train_test_split(X, y, test_size=0.33)

/usr/local/lib/python3.7/dist-packages/ipykernel_launcher.py:3: FutureWarning: In a future version of pandas all arguments of DataFrame.drop except for the argument 'labels' will be keyword-only
  This is separate from the ipykernel package so we can avoid doing imports until


#Modeling

##Regresion Arbol de decisiones


In [ ]:
from sklearn import tree

clf = tree.DecisionTreeRegressor()
clf = clf.fit(X_train, y_train)
y_pred = clf.predict(X_test)

In [ ]:
clf.score(X_train,y_train)

0.653261237524128

In [ ]:
from sklearn.metrics import max_error
max_error(y_test, y_pred)

17.5030394142928

Con este modelo puede agregar 18 votos "Si" extras en su error maximo

In [ ]:
from sklearn.metrics import r2_score
r2_score(y_test, y_pred)

0.355199453848584

En este modelo tiene un 35% de variacion al resultado original

##Support vector machine

In [ ]:
from sklearn import svm
regr = svm.SVR()
regr.fit(X_train, y_train)
y_pred = regr.predict(X_test)

In [ ]:
regr.score(X_train,y_train)

0.38558991283906996

In [ ]:
from sklearn.metrics import max_error
max_error(y_test, y_pred)

19.299602555734932

Con este modelo puede agregar 19 votos "Si" extras en su error maximo

In [ ]:
from sklearn.metrics import r2_score
r2_score(y_test, y_pred)

0.4021229962599694

En este modelo tiene un 40% de variacion al resultado original

##Stochastic Gradient Descent

In [ ]:
from sklearn.linear_model import SGDRegressor
from sklearn.pipeline import make_pipeline
from sklearn.preprocessing import StandardScaler

reg = make_pipeline(StandardScaler(),SGDRegressor(max_iter=1000, tol=1e-3))
reg.fit(X_train, y_train)
y_pred = reg.predict(X_test)

In [ ]:
reg.score(X_train,y_train)

0.4793040437815722

In [ ]:
from sklearn.metrics import max_error
max_error(y_test, y_pred)

16.1735342139205

Con este modelo puede agregar 16 votos "Si" extras en su error maximo

In [ ]:
from sklearn.metrics import r2_score
r2_score(y_test, y_pred)

0.5449416738103412

En este modelo tiene un 54% de variacion al resultado original

##PLSRegression

In [ ]:
from sklearn.cross_decomposition import PLSRegression
pls2 = PLSRegression(n_components=4)
pls2.fit(X_train, y_train)
PLSRegression()
y_pred = pls2.predict(X_test)

In [ ]:
pls2.score(X_train,y_train)

0.4793055600213577

In [ ]:
from sklearn.metrics import max_error
max_error(y_test, y_pred)

16.175556021382615

Con este modelo puede agregar 16 votos "Si" extras en su error maximo

In [ ]:
from sklearn.metrics import r2_score
r2_score(y_test, y_pred)

0.545066372784861

En este modelo tiene un 54% de variacion al resultado original

##Neural network models

In [ ]:
from sklearn.neural_network import MLPRegressor
from sklearn.datasets import make_regression

nnm = MLPRegressor(random_state=1, max_iter=500).fit(X_train, y_train)

y_pred= nnm.predict(X_test)

/usr/local/lib/python3.7/dist-packages/sklearn/neural_network/_multilayer_perceptron.py:696: ConvergenceWarning: Stochastic Optimizer: Maximum iterations (500) reached and the optimization hasn't converged yet.
  ConvergenceWarning,


In [ ]:
nnm.score(X_train,y_train)

0.3181330258518591

In [ ]:
from sklearn.metrics import max_error
max_error(y_test, y_pred)

17.5030394142928

Con este modelo puede agregar 18 votos "Si" extras en su error maximo

In [ ]:
from sklearn.metrics import r2_score
r2_score(y_test, y_pred)

0.3948238148967307

En este modelo tiene un 39% de variacion al resultado original